server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -494,7 +494,7 @@ private static boolean isTaskCancelledException(Exception e) {
     protected void onShardResult(Result result) {
         assert result.getShardIndex() != -1 : "shard index is not set";
         assert result.getSearchShardTarget() != null : "search shard target must not be null";
-        hasShardResponse.set(true);
+        hasShardResponse.compareAndExchangeRelease(false, true);
```

Comment:
Can you please document why acquire semantics are not needed here?

Reply:
Hmm same question here, does it really make sense to document this? We have no data dependency between this field and any other field and we only go false -> true here so if we do a plain read that's good enough, a potentially redundant write is still an improvement of the previous version. Maybe we should come up with a general list of what to use for what instead?